### ISSUE TITLE
link error on BatchNormImplBase<D, Derived>::pretty_print  when using icpc 

### ISSUE THREAD

**Issue by shlongsas** [pytorch/pytorch#58591]
## 🐛 Bug

The file torch/csrc/api/src/nn/modules/batchnorm.cpp includes a template class member function as below:

`template <size_t D, typename Derived>
void BatchNormImplBase<D, Derived>::pretty_print(std::ostream& stream) const {
  stream << std::boolalpha
         << "torch::nn::BatchNorm" << D << "d("
         << this->options.num_features() << ", "
         << "eps=" << this->options.eps() << ", "
         << "momentum=" << this->options.momentum().value() << ", "
         << "affine=" << this->options.affine() << ", "
         << "track_running_stats=" << this->options.track_running_stats() << ")";
}`

Other members of the class BatchNormImplBase are defined in .h file, and only this pretty_print function is defined in a .cpp file.

Whenever I use BatchNorm layer, it never gets instantiated properly, although I did not call this function in my application.  g++/gcc seems to ignore it, but icpc linker recognizes this as a link error.
